smcontroller/smctrlserver.go

```diff
@@ -0,0 +1,189 @@
+
+type smCtrlServer struct {
+	pb.UnimplementedSMServiceServer
+
+	messageSender          MessageSender
+	alertSender            AlertSender
+	monitoringSender       MonitoringSender
+	url                    string
+	grpcServer             *grpc.Server
+	listener               net.Listener
+	controllerCh           chan<- smCtrlInternalMsg
+	runStatusCh            chan<- launcher.NodeRunInstanceStatus
+	updateInstanceStatusCh chan<- []cloudprotocol.InstanceStatus
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// NewServer create update controller server.
+func newSmServer(cfg *config.Config, messageSender MessageSender, alertSender AlertSender,
+	monitoringSender MonitoringSender, controllerCh chan<- smCtrlInternalMsg,
+	runStatCh chan<- launcher.NodeRunInstanceStatus, updateInstanceStatusCh chan<- []cloudprotocol.InstanceStatus,
+	certProvider CertificateProvider, cryptcoxontext *cryptutils.CryptoContext, insecure bool,
+) (server *smCtrlServer, err error) {
+	log.WithField("host", cfg.SMController.SMServerURL).Debug("Start SM server")
+
+	server = &smCtrlServer{
+		messageSender: messageSender, alertSender: alertSender, monitoringSender: monitoringSender,
+		controllerCh: controllerCh, runStatusCh: runStatCh, updateInstanceStatusCh: updateInstanceStatusCh,
+	}
+
+	var opts []grpc.ServerOption
+
+	if !insecure {
+		certURL, keyURL, err := certProvider.GetCertificate(cfg.CertStorage, nil, "")
+		if err != nil {
+			return nil, aoserrors.Wrap(err)
+		}
+
+		tlsConfig, err := cryptcoxontext.GetClientMutualTLSConfig(certURL, keyURL)
+		if err != nil {
+			return nil, aoserrors.Wrap(err)
+		}
+
+		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	} else {
+		log.Info("GRPC server starts in insecure mode")
+	}
+
+	server.url = cfg.SMController.SMServerURL
+
+	server.grpcServer = grpc.NewServer(opts...)
+
+	pb.RegisterSMServiceServer(server.grpcServer, server)
+
+	return server, nil
+}
+
+// Start start update controller server.
+func (server *smCtrlServer) Start() (err error) {
+	server.listener, err = net.Listen("tcp", server.url)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	go func() {
+		if err := server.grpcServer.Serve(server.listener); err != nil {
+			log.Errorf("Can't serve gRPC server: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop stop update controller server.
+func (server *smCtrlServer) Stop() {
+	if server.grpcServer != nil {
+		server.grpcServer.Stop()
+	}
+
+	if server.listener != nil {
+		server.listener.Close()
+	}
+}
+
+// RegisterSM  register new SM client connection.
+func (server *smCtrlServer) RegisterSM(stream pb.SMService_RegisterSMServer) error {
+	message, err := stream.Recv()
+	if errors.Is(err, io.EOF) {
+		log.Warn("Unexpected end of UM stream")
```

Comment:
On close it will be received, so I guess it is expected?

Reply:
actually here we are waiting for the configuration to register client.
Connection handling implemented in smhandler

Reply:
Then why nil is returned in this case? It is error, right?

Reply:
in any error retun err